File: 2022/day-2.py
import asyncio
from util import main


def rock_paper_scissors(n: str, m: str) -> int:
    if m == "X":
        match n:
            case "A":
                return 1 + 3
            case "B":
                return 1
            case "C":
                return 1 + 6
            case _:
                pass
    elif m == "Y":
        match n:
            case "A":
                return 2 + 6
            case "B":
                return 2 + 3
            case "C":
                return 2
            case _:
                pass
    elif m == "Z":
        match n:
            case "A":
                return 3
            case "B":
                return 3 + 6
            case "C":
                return 3 + 3
            case _:
                pass
    return 0


def part_one(input: str) -> int:
    points = 0
    for line in input.split("\n"):
        arr = line.split()
        points += rock_paper_scissors(arr[0], arr[1])
    return points


def part_two(input: str) -> int:
    points = 0
    for line in input.split("\n"):
        arr = line.split()
        if arr[1] == "X":
            match arr[0]:
                case "A":
                    points += 3
                case "B":
                    points += 1
                case "C":
                    points += 2
                case _:
                    pass
        elif arr[1] == "Y":
            points += 3
            match arr[0]:
                case "A":
                    points += 1
                case "B":
                    points += 2
                case "C":
                    points += 3
                case _:
                    pass
        elif arr[1] == "Z":
            points += 6
            match arr[0]:
                case "A":
                    points += 2
                case "B":
                    points += 3
                case "C":
                    points += 1
                case _:
                    pass
    return points


asyncio.run(main(2022, 2, part_one, part_two))

File: 2025/solutions/day2/day2.go
package day2

import (
	"fmt"
	"log"
	"math"
	"strconv"
	"strings"

	mapset "github.com/deckarep/golang-set/v2"
)

func parseInput(input string) [][2]int {
	var parsed [][2]int
	lines := strings.SplitSeq(input, ",")
	for line := range lines {
		split := strings.Split(line, "-")
		start, err_1 := strconv.Atoi(split[0])
		end, err_2 := strconv.Atoi(split[1])
		if err_1 != nil || err_2 != nil {
			log.Fatal("Error in parsing input!")
		}
		parsed = append(parsed, [2]int{start, end})
	}
	return parsed
}

func invalidInRange(start int, end int, n int) []int {
	half_to_invalid := func(chunk int) int {
		invalid_str := strconv.Itoa(chunk)
		for range n - 1 {
			invalid_str = fmt.Sprintf("%s%d", invalid_str, chunk)
		}
		invalid, err := strconv.Atoi(invalid_str)
		if err != nil {
			log.Fatal("Fatal error!")
		}
		return invalid
	}

	var invalids []int
	var curr_chunk int
	start_str := strconv.Itoa(start)
	if len(start_str)%n != 0 {
		curr_chunk = int(math.Pow10(len(start_str) / n))
	} else {
		var err error
		curr_chunk, err = strconv.Atoi(start_str[0 : len(start_str)/n])
		if err != nil {
			log.Fatal("Fatal error!")
		}
	}
	curr_invalid := half_to_invalid(curr_chunk)
	for curr_invalid <= end {
		if curr_invalid >= start {
			invalids = append(invalids, curr_invalid)
		}
		curr_chunk += 1
		curr_invalid = half_to_invalid(curr_chunk)
	}
	return invalids
}

func partOne(input string) string {
	parsed := parseInput(input)
	invalid_sum := 0
	for _, id_range := range parsed {
		start, end := id_range[0], id_range[1]
		invalid_ids := invalidInRange(start, end, 2)
		for _, id := range invalid_ids {
			invalid_sum += id
		}
	}
	return strconv.Itoa(invalid_sum)
}

func partTwo(input string) string {
	parsed := parseInput(input)
	invalid_sum := 0
	for _, id_range := range parsed {
		start, end := id_range[0], id_range[1]
		invalid_ids := mapset.NewSet[int]()
		for n := 2; n <= len(strconv.Itoa(end)); n++ {
			invalid_ids.Append(invalidInRange(start, end, n)...)
		}
		for _, id := range invalid_ids.ToSlice() {
			invalid_sum += id
		}
	}
	return strconv.Itoa(invalid_sum)
}

func Puzzles() (func(string) string, func(string) string) {
	return partOne, partTwo
}
